cli/dcoscli/node/main.py

```diff
@@ -65,12 +65,214 @@ def _cmds():
+    """
+    Manage snapshots
+
+    :param list_snapshots: a list of available snapshots
+    :type  list_snapshots: bool
+    :param status: show snapshot job status
+    :type  status: bool
+    :param cancel: cancel snapshot job
+    :type  cancel: bool
+    :param snapshot_download: download a snapshot
+    :type  snapshot_download: string
+    :return: process return code
+    :type: int
+    """
+    if list_snapshots:
+        snapshot_response = _snapshot_get(
+            '/system/health/v1/report/snapshot/list/all')
+        available_snapshots = []
+        for _, snapshot_files in snapshot_response.items():
+            if snapshot_files is None:
+                continue
+            available_snapshots += map(lambda x: os.path.basename(x),
+                                       snapshot_files)
+
+        if not available_snapshots:
+            emitter.publish("No snapshots")
+            return 0
+        emitter.publish("Available snapshots:")
+        emitter.publish(', '.join(reversed(sorted(available_snapshots))))
+
+        return 0
+    elif status:
+        snapshot_response = _snapshot_get(
+            '/system/health/v1/report/snapshot/status/all')
+        for host, props in sorted(snapshot_response.items()):
+            emitter.publish(host)
+            for key, value in sorted(props.items()):
+                emitter.publish('  {}: {}'.format(key, value))
+            emitter.publish('\n')
+        return 0
+    elif cancel:
+        snapshot_response = _snapshot_post(
+            '/system/health/v1/report/snapshot/cancel')
+        emitter.publish(snapshot_response['status'])
+        return 0
+
+    # if we hit here, then the user wants to download the snapshot.
+    snapshot_response = _snapshot_download(snapshot_download)
+    emitter.publish(
+        'snapshot downloaded to {}'.format(
+            snapshot_response['snapshot_location']))
+    return 0
+
+
+def _do_request(url, method, payload=None, timeout=3, stream=False):
+    """
+    make HTTP request
+
+    :param url: url
+    :type url: string
+    :param method: HTTP method, GET or POST
+    :type  method: string
+    :param payload: json parameters for requests lib
+    :type  dict
+    :param timeout: HTTP request timeout, default 3 seconds
+    :type  timeout: integer
+    :param stream: stream parameter for requests lib
+    :type  stream: bool
+    :return: http response
+    :rtype: requests.Response
+    """
+    def _is_success(status_code):
+        # consider 400 and 503 to be successful status codes.
+        # API will return the error message.
+        if status_code in [200, 400, 503]:
+            return True
+        return False
+
+    do_request = {'GET': http.get, 'POST': http.post}.get(method)
+
+    # POST to snapshot api
+    client = mesos.DCOSClient()
+    url = client.get_dcos_url(url)
+    http_response = do_request(url, json=payload, is_success=_is_success,
+                               timeout=timeout, stream=stream)
```

Comment:
can we explicitly call `http.get` or `http.post` instead of do_request? As is, its hard to follow the code. This also allows us to get rid of `_snapshot_get|post` functions.